bluetooth: host: smp: Add bondable flag overlay per connection

mkapala-nordic opened the pull request:
The current API for changing the bondable mode uses the global flag. With Zephyr support for multiple Bluetooth identities, the API for changing the bondable mode should be more fine-grained. The bondable requirements of one identity should not have an impact on another identity which can have a different set of requirements. This change introduces function to overlay bondable flag per connection.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/57070

mkapala-nordic committed:
bluetooth: host: smp: Add bondable flag overlay per connection

The current API for changing the bondable mode uses the global flag.
With Zephyr support for multiple Bluetooth identities, the API for
changing the bondable mode should be more fine-grained.
The bondable requirements of one identity should not have an impact on
another identity which can have a different set of requirements.
This change introduces function to overlay bondable flag per
connection.

Signed-off-by: Mateusz Kapala <mateusz.kapala@nordicsemi.no>

mkapala-nordic commented:
CC: @kapi-no @MarekPieta

MarekPieta commented on subsys/bluetooth/host/smp.c:
```diff
@@ -288,6 +291,11 @@ static K_SEM_DEFINE(sc_local_pkey_ready, 0, 1);
  */
 #define BT_SMP_AUTH_CB_UNINITIALIZED	((atomic_ptr_val_t)bt_smp_pool)
 
+/* Value used to mark that bondable flag is not initialized.
```
per-connection bondable flag?

kapi-no reviewed (approved):
Looking good (apart from small doc improvements suggested by @MarekPieta)

kapi-no commented:
The naming convention and the implementation for this new `bt_bondable_overlay` API are similar to the `bt_smp_auth_cb_overlay` API function.

mkapala-nordic commented:
Adressed @MarekPieta comments.

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1081,6 +1081,25 @@ void bt_conn_cb_register(struct bt_conn_cb *cb);
+ *  @param conn Connection object.
+ *  @param enable Value allowing/disallowing to be bondable.
+ */
+int bt_bondable_overlay(struct bt_conn *conn, bool enable);
```
I feel like `bt_conn_set_bondable()` would be the more appropriate name. I know we have `bt_smp_auth_cb_overlay()` as an internal API, but public APIs operating on a `struct bt_conn *` should preferably be prefixed with `bt_conn_`.

mkapala-nordic replied:
Agreed, name changed.

jori-nordic reviewed (changes requested):
could you add a bsim test checking the new functionality?
and also mark this new API as experimental (in case we later find some better way to achieve the same thing)

mkapala-nordic committed:
bluetooth: tests: Add bsim test for setting bondable flag per-conn

Added test for the bt_conn_set_bondable API function.
Check if we can pair without setting the bonding flag on the
per-connection basis if the device was already bonded on the
other identity.

Signed-off-by: Mateusz Kapala <mateusz.kapala@nordicsemi.no>

mkapala-nordic commented:
Marked as experimental + added bsim test as per @jori-nordic request.

kapi-no commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -600,6 +600,13 @@ config BT_BONDING_REQUIRED
+	bool "Set/clear the bonding flag per-connection [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  Enable support for setting/clearing the bonding flag on a
+	  per-connection basis.
```
```suggestion
	  Enable support for the bt_conn_set_bondable API function that is 
	  used to set/clear the bonding flag on a per-connection basis.
```

MarekPieta reviewed (approved):
Change in host looks good

jori-nordic commented on tests/bsim/bluetooth/host/security/bond_per_connection/src/bs_bt_utils.c:
```diff
@@ -0,0 +1,191 @@
+
+	SET_FLAG(flag_is_connected);
+
+	if (bt_conn_set_bondable(conn, bondable)) {
```
could you add another round of bonding without calling `bt_conn_set_bondable`?
like have a bondable by default device, try bonding and see that it works and then continue with the two test rounds you already have

mkapala-nordic replied:
Added such case.

carlescufi commented:
@jhedberg would you mind taking another look?
